cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java

```diff
@@ -71,11 +94,49 @@ public HBaseMetricsTable(DatasetContext datasetContext, DatasetSpecification spe
     this.columnFamily = TableProperties.getColumnFamilyBytes(spec.getProperties());
   }
 
+  private void initializeV3Vars(DatasetSpecification spec) {
+    boolean isV3Table = spec.getName().contains("v3");
+    this.scanExecutor = null;
+    this.rowKeyDistributor = null;
+    if (isV3Table) {
+      RejectedExecutionHandler callerRunsPolicy = new RejectedExecutionHandler() {
+
+        @Override
+        public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+          REJECTION_LOG.info(
+            "No more threads in the HBase scan thread pool. Consider increase {}. Scan from caller thread {}",
+            Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS, Thread.currentThread().getName()
+          );
+          // Runs it from the caller thread
+          if (!executor.isShutdown()) {
+            r.run();
+          }
+        }
+      };
+
+      int maxScanThread = spec.getIntProperty(Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS, 96);
+      // Creates a executor that will shrink to 0 threads if left idle
+      // Uses daemon thread, hence no need to worry about shutdown
+      // When all threads are busy, use the caller thread to execute
+      this.scanExecutor = new ThreadPoolExecutor(0, maxScanThread, 60L, TimeUnit.SECONDS,
+                                                 new SynchronousQueue<Runnable>(),
+                                                 Threads.createDaemonThreadFactory("metrics-hbase-scanner-%d"),
+                                                 callerRunsPolicy);
+
+      int splits = spec.getIntProperty(Constants.Metrics.METICS_HBASE_TABLE_SPLITS,
```

Comment:
So storing this in the dataset properties and retrieving it here means that it is not hard coded? What happens if it gets updated , say from 16 to 32? Does it make sense to have it as a property?